jena-arq/src/main/java/org/apache/jena/sparql/engine/main/VarFinder.java

```diff
@@ -351,9 +389,26 @@ public void visit(OpProject opProject) {
             assignMentions.addAll(subUsage.assignMentions);
         }
 
+
         @Override
         public void visit(OpTable opTable) {
-            defines.addAll(opTable.getTable().getVars());
+            Table table = opTable.getTable();
+            Set<Var> definedVars = new HashSet<>(table.getVars());
+            List<Var> optVars = new ArrayList<>();
+            Iterator<Binding> rowIt = table.rows();
+            while (rowIt.hasNext()) {
+                Binding row = rowIt.next();
+                Iterator<Var> varIt = definedVars.iterator();
+                while (varIt.hasNext()) {
+                    Var var = varIt.next();
+                    if (!row.contains(var)) {
```

Comment:
This seems overly aggressive at first glance.  If I am reading this correctly any row that does not contain a value triggers the join linearisation being blocked. What happens if you have multiple rows with some values and some `UNDEF`, is that going to create bad results if we allow linearisation?

Perhaps a different solution is needed around handling of linear joins for `OpTable` to just never substitute over an `UNDEF` though guess that might be significantly more involved?

Reply:
> If I am reading this correctly any row that does not contain a value triggers the join linearisation being blocked.

Its per variable - so if there is an UNDEF value for a variable then linearization for that variable gets blocked.
Conversely, if the variable is defined in all rows its added to the defined set.
I was creating test cases for skewed multi-variable joins (i.e. joins on multiple variables where some of them are unbound) when I noticed that I get incorrect results.